heron/stmgr/src/cpp/manager/stmgr.cpp

```diff
@@ -795,44 +835,113 @@ void StMgr::SendStartBackPressureToOtherStMgrs() {
 void StMgr::SendStopBackPressureToOtherStMgrs() { clientmgr_->SendStopBackPressureToOtherStMgrs(); }
 
 // Do any actions if a stmgr client connection dies
-void StMgr::HandleDeadStMgrConnection(const sp_string&) {
-  // TODO(srkukarni) For Stateful Topologies, we need to do a restore
+void StMgr::HandleDeadStMgrConnection(const sp_string& _stmgr_id) {
+  // If we are stateful topology, we need to send a resetTopology message
+  // in case we are not in 2pc
+  if (stateful_restorer_) {
+    if (!stateful_restorer_->InProgress() && tmaster_client_) {
+      LOG(INFO) << "We lost connection withi stmgr " << _stmgr_id
+                << " and hence sending ResetTopology message to tmaster";
+      tmaster_client_->SendResetTopologyState(_stmgr_id, -1, "Dead Stmgr");
+      restore_initiated_metrics_->incr();
+    } else {
+      // We are in restore
+      stateful_restorer_->HandleDeadStMgrConnection();
+    }
+  }
 }
 
 void StMgr::HandleAllStMgrClientsRegistered() {
   // If we are stateful topology, we might want to continue our restore process
-  // TODO(srkukarni) Complete this
+  if (stateful_restorer_) {
+    stateful_restorer_->HandleAllStMgrClientsConnected();
+  }
 }
 
 void StMgr::HandleAllInstancesConnected() {
-  // Now we can connect to the tmaster
-  StartTMasterClient();
+  // StmgrServer told us that all instances are connected to us
+  if (stateful_restorer_) {
+    if (stateful_restorer_->InProgress()) {
+      // We are in the middle of a restore
+      stateful_restorer_->HandleAllInstancesConnected();
+    } else if (tmaster_client_ && tmaster_client_->IsConnected()) {
```

Comment:
could you elaborate why in this case the `ResetTopologyState` message needs to be sent?